doc: add documentation for testbeds (#347)
danth/stylix: docs/src/testbeds.md

- Reducing the risk of breaking your system while reviewing pull requests.

Testbeds are also built by GitHub Actions for every pull request. This is less
beneficial compared to running them yourself, since it cannot visually check
the theme, however it can catch build failures which may have been missed
otherwise.

## Creation

New testbeds are defined by creating a file called `testbed.nix` within the
folder for the corresponding target. This file will automatically be loaded
as a NixOS module, with options such as `stylix.image` already defined.
The module should include any options necessary to install the target and
any supporting software - for example, a window manager.

If the target can only be used through Home Manager, you can write a
Home Manager module within the NixOS module using the following format:

```nix
{
    home-manager.sharedModules = [{
        # Write Home Manager options here
    }];
}
```

Using `home-manager.sharedModules` is preferred over `home-manager.users.guest`
since it allows us to easily change the username or add additional users in
the future.

Once the module is complete, use `git add` to track the file, then the new

> packages will be available to use with `nix run` as described below.

Author: In personal experience I've often seen people run into things not working because they didn't run `git add`, so Nix didn't see the new file. This is even more confusing when we have automatic imports, since a file which wasn't added will simply not be imported rather than causing an error that the import doesn't exist.

As it's only one sentence I think it's worth keeping in.

> packages will be available to use with `nix run` as described below.

Author: How about simply `[available to use](#usage)`? Since the usage section describes the exact command.

> 
> Once the module is complete, use `git add` to track the file, then the new
> packages will be available to use with `nix run` as described below.

Reviewer: > How about simply `[available to use](#usage)`? Since the usage section describes the exact command.

Sounds good. We could also do the following. Although, I don't think those differences are that important.

```suggestion

Once the module is complete, use `git add` to track the file, and then the new
packages will be [available](#usage).
```

> packages will be available to use with `nix run` as described below.

Author: OK, I went for `[available to use](#usage)`.

## Usage

You can list the available testbeds by running this command from anywhere
within the repository:

```sh
nix flake show
```

Which prints something like:

```
github:danth/stylix
└───packages
    └───x86_64-linux
        ├───docs: package 'stylix-book'
        ├───palette-generator: package 'palette-generator'
        ├───testbed-gnome-dark: package 'testbed-gnome-dark'
        ├───testbed-gnome-light: package 'testbed-gnome-light'
        ├───testbed-kde-dark: package 'testbed-kde-dark'
        └───testbed-kde-light: package 'testbed-kde-light'
```

(This has been edited down to only the relevant parts.)

> The packages starting with `testbed-` are the testbeds.

Author: I intended this not to be updated and to just show the current list as an example.

Given that the docs are meant to be compiled with `nix build .#docs` and viewed through the website, we could consider replacing this entire section with an automatically generated table, and avoid directing users to `nix flake show` at all.

Advantages of that new method:
- Always up to date
- Will error if the format is changed, so you're forced to keep the docs in sync

Disadvantages:
- The raw markdown is less readable
- Additional code to maintain
- Only shows the current state of `master` unless you build the docs locally
- If the list becomes very long we may need to move it to a separate page, which breaks the flow of the text

> The packages starting with `testbed-` are the testbeds.

Reviewer: > Given that the docs are meant to be compiled with `nix build .#docs` and viewed through the website, we could consider replacing this entire section with an automatically generated table, and avoid directing users to `nix flake show` at all.

I am not sure whether the added documentation generation complexity is worth it.

> The packages starting with `testbed-` are the testbeds.

Author: I agree, there are far more disadvantages than advantages on that list.

Note that the suggestion of `nix flake show | grep testbed` is a bit bugged since it outputs parts of the progress bar interleaved with the real output:

![screenshot](https://github.com/danth/stylix/assets/28959268/1d98c86e-129c-4e09-9d27-22dab016a7e1)

Whatever we need to do to work around this is probably too complex to include in the documentation without a full explanation, which distracts from the focus of the text.

You can select a testbed to run using:

> ```sh

Author: Looks like `console` is meant to include a prefix like `foo@bar:~$ ` before each command and then show the output below, within the same code block.

We can use that to eliminate the "Which prints something like" line between the command and its output.

> ```sh

Reviewer: Is the following simpler format not valid:

```console
$ seq 3
1
2
3
```

> nix run .#testbed-gnome-dark

Author: Consider also using `«` and `»` characters (right alt + z and right alt + x respectively)

- Consistent with the format used in NixOS option docs
- Clearer that it's not part of the real syntax

> nix run .#testbed-gnome-dark

Reviewer: > Consider also using `«` and `»` characters (right alt + z and right alt + x respectively)
> 
> * Consistent with the format used in NixOS option docs
> * Clearer that it's not part of the real syntax

My bad. I am used to the ASCII version.

```

This may take a few minutes to download and build the virtual machine.

Once it starts, it should open a window similar to the screenshot below. The
contents of the the virtual machine will vary depending on the target you
selected earlier.

![GDM login screen with a dark background color and showing a guest user](testbed-gnome-dark.png)

If the testbed includes a login screen, the guest user should log in
automatically when selected. Depending on the software used, you may still be
presented with a password prompt - in which case you can leave it blank and

> just press enter.

Author: IMO there is no need for this to be in a monospace font, since it isn't code.

However, I tweaked wording of the sentence to be more clear about what to do.
